File: web/packages/dtx-core/tests/helpers/memory-fs.ts
import {
  decodeTextWithBom,
  type DirEntry,
  type FileSystemBackend,
} from '../../src/scanner/fs-backend.js';

type Entry = { kind: 'file'; bytes: Uint8Array } | { kind: 'dir' };

/**
 * Test-only in-memory FileSystemBackend. Paths are POSIX; directories are
 * implicit (any path that is a prefix of a file path acts as a directory).
 */
export class MemoryFs implements FileSystemBackend {
  private readonly entries = new Map<string, Entry>();

  setFile(path: string, content: string | Uint8Array): void {
    const bytes = typeof content === 'string' ? new TextEncoder().encode(content) : content;
    this.entries.set(normalize(path), { kind: 'file', bytes });
    // Register parent chain as directories for exists()/listDir().
    let parent = parentOf(normalize(path));
    while (parent.length > 0 && !this.entries.has(parent)) {
      this.entries.set(parent, { kind: 'dir' });
      parent = parentOf(parent);
    }
  }

  async listDir(path: string): Promise<DirEntry[]> {
    const norm = normalize(path);
    const prefix = norm === '' ? '' : norm + '/';
    const seen = new Map<string, DirEntry>();
    for (const [p, entry] of this.entries) {
      if (p === norm) continue;
      if (!p.startsWith(prefix)) continue;
      const rest = p.slice(prefix.length);
      const name = rest.split('/')[0]!;
      const childPath = prefix + name;
      if (seen.has(name)) continue;
      const childEntry = this.entries.get(childPath);
      const isDir = childEntry?.kind === 'dir';
      seen.set(name, {
        name,
        path: childPath,
        isDirectory: isDir,
        isFile: !isDir,
      });
    }
    return Array.from(seen.values());
  }

  async readFile(path: string): Promise<ArrayBuffer> {
    const e = this.entries.get(normalize(path));
    if (!e || e.kind !== 'file') throw new Error(`not a file: ${path}`);
    return e.bytes.slice().buffer;
  }

  async readText(path: string, encoding: string = 'shift-jis'): Promise<string> {
    const e = this.entries.get(normalize(path));
    if (!e || e.kind !== 'file') throw new Error(`not a file: ${path}`);
    return decodeTextWithBom(e.bytes.buffer.slice(e.bytes.byteOffset, e.bytes.byteOffset + e.bytes.byteLength) as ArrayBuffer, encoding);
  }

  async exists(path: string): Promise<boolean> {
    return this.entries.has(normalize(path));
  }
}

function normalize(path: string): string {
  return path.replace(/^\/+/, '').replace(/\/+/g, '/').replace(/\/+$/, '');
}

function parentOf(path: string): string {
  const idx = path.lastIndexOf('/');
  return idx < 0 ? '' : path.slice(0, idx);
}

File: native/quest-spatial/app/src/test/kotlin/com/dtxmania/quest/dtxcore/scanner/MemoryFs.kt
package com.dtxmania.quest.dtxcore.scanner

/**
 * Test-only in-memory [FileSystemBackend]. Paths are POSIX; directories
 * are implicit (any path that is a prefix of a file path acts as a
 * directory).
 *
 * Ported from `web/packages/dtx-core/tests/helpers/memory-fs.ts`. Tests
 * may also override [listDir] / [readText] / [exists] dynamically by
 * reassigning the matching `var`s — used by the "reports errors instead
 * of throwing" case to inject a synthetic failure.
 */
class MemoryFs : FileSystemBackend {

    private sealed class Entry {
        class File(val bytes: ByteArray) : Entry()
        object Dir : Entry()
    }

    private val entries = LinkedHashMap<String, Entry>()

    /** Add or overwrite a file. */
    fun setFile(path: String, content: String) {
        setFile(path, content.toByteArray(Charsets.UTF_8))
    }

    fun setFile(path: String, content: ByteArray) {
        val norm = normalize(path)
        entries[norm] = Entry.File(content)
        // Register parent chain as directories for exists()/listDir().
        var parent = parentOf(norm)
        while (parent.isNotEmpty() && !entries.containsKey(parent)) {
            entries[parent] = Entry.Dir
            parent = parentOf(parent)
        }
    }

    /** Hook for test cases that need to inject a failing listDir on a
     *  specific path. Override callbacks may delegate back to the real
     *  implementation via [listDirImpl]. */
    var listDirOverride: ((String) -> List<DirEntry>)? = null

    override fun listDir(path: String): List<DirEntry> {
        listDirOverride?.let { return it(path) }
        return listDirImpl(path)
    }

    fun listDirImpl(path: String): List<DirEntry> {
        val norm = normalize(path)
        val prefix = if (norm.isEmpty()) "" else "$norm/"
        val seen = LinkedHashMap<String, DirEntry>()
        for ((p, entry) in entries) {
            if (p == norm) continue
            if (!p.startsWith(prefix)) continue
            val rest = p.substring(prefix.length)
            val name = rest.substringBefore('/')
            if (seen.containsKey(name)) continue
            val childPath = prefix + name
            val childEntry = entries[childPath]
            val isDir = childEntry is Entry.Dir
            seen[name] = DirEntry(
                name = name,
                path = childPath,
                isDirectory = isDir,
                isFile = !isDir,
            )
        }
        return seen.values.toList()
    }

    override fun readFile(path: String): ByteArray {
        val e = entries[normalize(path)]
        if (e !is Entry.File) throw IllegalStateException("not a file: $path")
        return e.bytes.copyOf()
    }

    override fun readText(path: String, encoding: String): String {
        val e = entries[normalize(path)]
        if (e !is Entry.File) throw IllegalStateException("not a file: $path")
        return decodeTextWithBom(e.bytes, encoding)
    }

    override fun exists(path: String): Boolean = entries.containsKey(normalize(path))

    private fun normalize(path: String): String =
        path.trim('/').replace(Regex("""/+"""), "/")

    private fun parentOf(path: String): String {
        val idx = path.lastIndexOf('/')
        return if (idx < 0) "" else path.substring(0, idx)
    }
}

/** Convenience: build a [MemoryFs] from a `path → text-content` map. */
fun makeFs(files: Map<String, String>): MemoryFs {
    val fs = MemoryFs()
    for ((path, content) in files) fs.setFile(path, content)
    return fs
}
